File: examples/spacebrew_skeleton_receiver_JS/js/sbxSkeleton.js
/**
 * 
 * sbxSkeleton: Spacebrew extension for sending and receiving openNI skeleton types via spacebrew
 * --------------------------------
 * 
 * @author 		Bernardo Schorr
 * @date 		May 31, 2014
 * 
 */

//sbxSkeleton main function. Will hold the positions for each of the openNI joints received.
function skeleton () {

			this.head = {x :0, y: 0};
			this.neck = {x :0, y: 0};
			this.torso = {x :0, y: 0};

			this.left_shoulder = {x :0, y: 0};
			this.left_elbow = {x :0, y: 0};
			this.left_hand = {x :0, y: 0};

			this.right_shoulder = {x :0, y: 0};
			this.right_elbow = {x :0, y: 0};
			this.right_hand = {x :0, y: 0};

			this.left_hip = {x :0, y: 0};
			this.left_knee = {x :0, y: 0};
			this.left_foot = {x :0, y: 0};

			this.right_hip = {x :0, y: 0};
			this.right_knee = {x :0, y: 0};
			this.right_foot = {x :0, y: 0};

			this.update = function( incomingSkeleton ) {

				this.head = incomingSkeleton.head;
				this.neck = incomingSkeleton.neck;
				this.torso = incomingSkeleton.torso;
				
				this.left_shoulder = incomingSkeleton.left_shoulder;
				this.left_elbow = incomingSkeleton.left_elbow;
				this.left_hand = incomingSkeleton.left_hand;

				this.right_shoulder = incomingSkeleton.right_shoulder;
				this.right_elbow = incomingSkeleton.right_elbow;
				this.right_hand = incomingSkeleton.right_hand;

				this.left_hip = incomingSkeleton.left_hip;
				this.left_knee = incomingSkeleton.left_knee;
				this.left_foot = incomingSkeleton.left_foot;

				this.right_hip = incomingSkeleton.right_hip;
				this.right_knee = incomingSkeleton.right_knee;
				this.right_foot = incomingSkeleton.right_foot;
			}

			this.draw = function ( ctx ) {

				var jointSize = 10;

				//draw the nodes
				ctx.fillRect (this.head.x-jointSize/2, this.head.y-jointSize/2, jointSize, jointSize);
				ctx.fillRect (this.neck.x-jointSize/2, this.neck.y-jointSize/2, jointSize, jointSize);
				ctx.fillRect (this.torso.x-jointSize/2, this.torso.y-jointSize/2, jointSize, jointSize);

				ctx.fillRect (this.left_shoulder.x-jointSize/2, this.left_shoulder.y-jointSize/2, jointSize, jointSize);
				ctx.fillRect (this.left_elbow.x-jointSize/2, this.left_elbow.y-jointSize/2, jointSize, jointSize);
				ctx.fillRect (this.left_hand.x-jointSize/2, this.left_hand.y-jointSize/2, jointSize, jointSize);

				ctx.fillRect (this.right_shoulder.x-jointSize/2, this.right_shoulder.y-jointSize/2, jointSize, jointSize);
				ctx.fillRect (this.right_elbow.x-jointSize/2, this.right_elbow.y-jointSize/2, jointSize, jointSize);
				ctx.fillRect (this.right_hand.x-jointSize/2, this.right_hand.y-jointSize/2, jointSize, jointSize);

				ctx.fillRect (this.left_hip.x-jointSize/2, this.left_hip.y-jointSize/2, jointSize, jointSize);
				ctx.fillRect (this.left_knee.x-jointSize/2, this.left_knee.y-jointSize/2, jointSize, jointSize);
				ctx.fillRect (this.left_foot.x-jointSize/2, this.left_foot.y-jointSize/2, jointSize, jointSize);

				ctx.fillRect (this.right_hip.x-jointSize/2, this.right_hip.y-jointSize/2, jointSize, jointSize);
				ctx.fillRect (this.right_knee.x-jointSize/2, this.right_knee.y-jointSize/2, jointSize, jointSize);
				ctx.fillRect (this.right_foot.x-jointSize/2, this.right_foot.y-jointSize/2, jointSize, jointSize);

				//draw the connections
				ctx.beginPath();
				
				//head to neck to torso
				ctx.moveTo (this.head.x, this.head.y);
				ctx.lineTo (this.neck.x, this.neck.y);

				ctx.moveTo (this.neck.x, this.neck.y);
				ctx.lineTo (this.torso.x, this.torso.y);

				//neck to shoulders
				ctx.moveTo (this.neck.x, this.neck.y);
				ctx.lineTo (this.left_shoulder.x, this.left_shoulder.y);

				ctx.moveTo (this.neck.x, this.neck.y);
				ctx.lineTo (this.right_shoulder.x, this.right_shoulder.y);

				//left arm
				ctx.moveTo (this.left_shoulder.x, this.left_shoulder.y);
				ctx.lineTo (this.left_elbow.x, this.left_elbow.y);

				ctx.moveTo (this.left_elbow.x, this.left_elbow.y);				
				ctx.lineTo (this.left_hand.x, this.left_hand.y);

				//right arm
				ctx.moveTo (this.right_shoulder.x, this.right_shoulder.y);
				ctx.lineTo (this.right_elbow.x, this.right_elbow.y);

				ctx.moveTo (this.right_elbow.x, this.right_elbow.y);				
				ctx.lineTo (this.right_hand.x, this.right_hand.y);

				//torso to knees
				ctx.moveTo (this.torso.x, this.torso.y);
				ctx.lineTo (this.left_hip.x, this.left_hip.y);

				ctx.moveTo (this.torso.x, this.torso.y);
				ctx.lineTo (this.right_hip.x, this.right_hip.y);

				//left leg
				ctx.moveTo (this.left_hip.x, this.left_hip.y);
				ctx.lineTo (this.left_knee.x, this.left_knee.y);

				ctx.moveTo (this.left_knee.x, this.left_knee.y);
				ctx.lineTo (this.left_foot.x, this.left_foot.y);

				//right leg
				ctx.moveTo (this.right_hip.x, this.right_hip.y);
				ctx.lineTo (this.right_knee.x, this.right_knee.y);

				ctx.moveTo (this.right_knee.x, this.right_knee.y);
				ctx.lineTo (this.right_foot.x, this.right_foot.y);

				ctx.stroke();
			}
		}

File: examples/spacebrew_skeleton_sender_OF/src/sbxSkeleton_OF/sbxSkeleton.cpp
//
//  sbxSkeleton.cpp
//
//  Created by Bernardo Santos Schorr on 3/31/14.
//
//

#include "sbxSkeleton.h"

namespace sbxSkeleton {

void skeletonSender::setup( ofxNiTE2::UserTracker &_tracker, Spacebrew::Connection  &_spacebrew, Spacebrew::Config &config, string _name ) {
    
    name = _name;
    config.addPublish(name, "skeleton", "");
    spacebrew = &_spacebrew;
    tracker = &_tracker;
    
}

void skeletonSender::sendSkeleton( ofxNiTE2::User::Ref user ) {
    
    //spacebrew sending data
    Json::Value skeleton = Json::Value( Json::objectValue );
    
    Json::Value neck = Json::Value( Json::objectValue );
    Json::Value head = Json::Value( Json::objectValue );
    
    Json::Value left_shoulder = Json::Value( Json::objectValue );
    Json::Value left_elbow = Json::Value( Json::objectValue );
    Json::Value left_hand = Json::Value( Json::objectValue );
    
    Json::Value right_shoulder = Json::Value( Json::objectValue );
    Json::Value right_elbow = Json::Value( Json::objectValue );
    Json::Value right_hand = Json::Value( Json::objectValue );
    
    Json::Value left_hip = Json::Value( Json::objectValue );
    Json::Value left_knee = Json::Value( Json::objectValue );
    Json::Value left_foot = Json::Value( Json::objectValue );
    
    Json::Value right_hip = Json::Value( Json::objectValue );
    Json::Value right_knee = Json::Value( Json::objectValue );
    Json::Value right_foot = Json::Value( Json::objectValue );
    
    Json::Value torso = Json::Value( Json::objectValue );
    
    //neck and head
    
    neck["x"] = round (getScreenPosition(user->getJoint(nite::JOINT_NECK)).x);
    neck["y"] = round (getScreenPosition(user->getJoint(nite::JOINT_NECK)).y);
    
    head["x"] = round (getScreenPosition(user->getJoint(nite::JOINT_HEAD)).x);
    head["y"] = round (getScreenPosition(user->getJoint(nite::JOINT_HEAD)).y);
    
    //left arm
    left_shoulder["x"] = round (getScreenPosition(user->getJoint(nite::JOINT_LEFT_SHOULDER)).x);
    left_shoulder["y"] = round (getScreenPosition(user->getJoint(nite::JOINT_LEFT_SHOULDER)).y);
    
    left_elbow["x"] = round (getScreenPosition(user->getJoint(nite::JOINT_LEFT_ELBOW)).x);
    left_elbow["y"] = round (getScreenPosition(user->getJoint(nite::JOINT_LEFT_ELBOW)).y);
    
    left_hand["x"] = round (getScreenPosition(user->getJoint(nite::JOINT_LEFT_HAND)).x);
    left_hand["y"] = round (getScreenPosition(user->getJoint(nite::JOINT_LEFT_HAND)).y);
    
    //right arm
    right_shoulder["x"] = round (getScreenPosition(user->getJoint(nite::JOINT_RIGHT_SHOULDER)).x);
    right_shoulder["y"] = round (getScreenPosition(user->getJoint(nite::JOINT_RIGHT_SHOULDER)).y);
    
    right_elbow["x"] = round (getScreenPosition(user->getJoint(nite::JOINT_RIGHT_ELBOW)).x);
    right_elbow["y"] = round (getScreenPosition(user->getJoint(nite::JOINT_RIGHT_ELBOW)).y);
    
    right_hand["x"] = round (getScreenPosition(user->getJoint(nite::JOINT_RIGHT_HAND)).x);
    right_hand["y"] = round (getScreenPosition(user->getJoint(nite::JOINT_RIGHT_HAND)).y);
    
    //left leg
    left_hip["x"] = round (getScreenPosition(user->getJoint(nite::JOINT_LEFT_HIP)).x);
    left_hip["y"] = round (getScreenPosition(user->getJoint(nite::JOINT_LEFT_HIP)).y);
    
    left_knee["x"] = round (getScreenPosition(user->getJoint(nite::JOINT_LEFT_KNEE)).x);
    left_knee["y"] = round (getScreenPosition(user->getJoint(nite::JOINT_LEFT_KNEE)).y);
    
    left_foot["x"] = round (getScreenPosition(user->getJoint(nite::JOINT_LEFT_FOOT)).x);
    left_foot["y"] = round (getScreenPosition(user->getJoint(nite::JOINT_LEFT_FOOT)).y);
    
    //right leg
    right_hip["x"] = round (getScreenPosition(user->getJoint(nite::JOINT_RIGHT_HIP)).x);
    right_hip["y"] = round (getScreenPosition(user->getJoint(nite::JOINT_RIGHT_HIP)).y);
    
    right_knee["x"] = round (getScreenPosition(user->getJoint(nite::JOINT_RIGHT_KNEE)).x);
    right_knee["y"] = round (getScreenPosition(user->getJoint(nite::JOINT_RIGHT_KNEE)).y);
    
    right_foot["x"] = round (getScreenPosition(user->getJoint(nite::JOINT_RIGHT_FOOT)).x);
    right_foot["y"] = round (getScreenPosition(user->getJoint(nite::JOINT_RIGHT_FOOT)).y);
    
    //torso
    torso["x"] = round (getScreenPosition(user->getJoint(nite::JOINT_TORSO)).x);
    torso["y"] = round (getScreenPosition(user->getJoint(nite::JOINT_TORSO)).y);
    
    //build skeleton
    skeleton["neck"] = neck;
    skeleton["head"] = head;
    skeleton["torso"] = torso;
    
    skeleton["left_shoulder"] = left_shoulder;
    skeleton["left_elbow"] = left_elbow;
    skeleton["left_hand"] = left_hand;
    
    skeleton["right_shoulder"] = right_shoulder;
    skeleton["right_elbow"] = right_elbow;
    skeleton["right_hand"] = right_hand;
    
    skeleton["left_hip"] = left_hip;
    skeleton["left_knee"] = left_knee;
    skeleton["left_foot"] = left_foot;
    
    skeleton["right_hip"] = right_hip;
    skeleton["right_knee"] = right_knee;
    skeleton["right_foot"] = right_foot;
    
    spacebrew->send( name, "skeleton", skeleton.toStyledString());
    
}
    ofVec3f skeletonSender::getScreenPosition( ofxNiTE2::Joint joint ) {
        
        ofVec3f pos;
        pos = joint.getGlobalPosition();
        pos = tracker->getOverlayCamera().worldToScreen(pos);
        
        return pos;
    }
}
